clippy_lints/src/path_buf_push_overwrite.rs

```diff
@@ -0,0 +1,65 @@
+use crate::utils::{match_type, paths, span_help_and_lint, walk_ptrs_ty};
+use if_chain::if_chain;
+use rustc::hir::*;
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_tool_lint, lint_array};
+use std::path::Path;
+use syntax::ast::LitKind;
+
+declare_clippy_lint! {
+    /// **What it does:*** Checks for `push` calls on `PathBuf` that can cause overwrites.
+    ///
+    /// **Why is this bad?** Calling `push` with a root path at the start can overwrite the
+    /// previous defined path.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```ignore
+    /// let mut x = PathBuf::from("/foo");
+    /// x.push("/bar");
+    /// println!("x: {}", x.display()); // "/bar
+    /// ```
+    /// Could be written:
+    ///
+    /// ```ignore
+    /// let mut x = PathBuf::from("/foo");
+    /// x.push("bar");
+    /// println!("x: {}", x.display()); // "/foo/bar
+    /// ```
+    pub PATH_BUF_PUSH_OVERWRITE,
+    style,
+    "calling `push` with file system root on `PathBuf` can overwrite it"
+}
+
+#[derive(Copy, Clone)]
+pub struct Pass;
+
+impl LintPass for Pass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(PATH_BUF_PUSH_OVERWRITE)
+    }
+
+    fn name(&self) -> &'static str {
+        "PathBufPushOverwrite"
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if_chain! {
+            if let ExprKind::MethodCall(ref path, _, ref args) = expr.node;
+            if path.ident.name == "push";
```

Comment:
Can you add a sanity check for the `args` vector after this line?
```suggestion
            if path.ident.name == "push";
            if args.len() == 2;
```
Otherwise the next line might panic